This notebook reads in files 'prices.csv' and 'stores.json' and reorganizes the data into a cross-tabulation of regional prices alongside each other, broken down by banner, then outputs the data to a spreadsheet.  The notebook also finds any anomolies in the data.

In [13]:
import pandas as pd
import json
import numpy as np

Read in json file 'stores.json' and put data into list of dictionaries called store_dict

In [2]:
with open('stores.json') as f:
    store_dicts = json.load(f)

Convert list of dictionaries to pandas dataframe named store_df

In [3]:
store_df = pd.DataFrame(store_dicts)

read in csv file with pricing data, 'prices.csv', and create pandas dataframe named prices_df

In [4]:
prices_df = pd.read_csv('prices.csv')

verify that the store data, store_df, contains all the information needed for our price data by creating a numpy array of all unique store ids in the prices dataframe and a numpy array of all unique store ids in the store dataframe and then comparing the two

In [8]:
uniq_stores = prices_df['Store ID'].unique()
all_stores = store_df['Store ID'].unique()

In [14]:
np.setdiff1d(uniq_stores, all_stores)

array([38472, 60957])

We find that there are two store ids, 38472 and 60957, in our pricing data for which we do not have Banner and Regional data. In the next step we will merge our price and store dataframes by store id as the intersection of keys from both frames, so the prices from the two stores for which we do not have info will simply not be included

In [15]:
merged = pd.merge(prices_df, store_df, on='Store ID')

Now we have a dataframe with prices and store info all together.

In [16]:
merged

Auditor ID      Date  Price  Store ID        UPC       Banner  \
0             234  10/18/17  24.95     66999  268588472      Walmart   
1             234  10/27/17  49.71     66999  475245085      Walmart   
2             234  10/20/17  25.75     66999  126967843      Walmart   
3             234  10/23/17  18.81     66999  708930835      Walmart   
4             234  10/23/17  33.32     66999  325885139      Walmart   
5             234  10/16/17  40.43     66999  509419735      Walmart   
6             234  10/22/17  35.10     66999  736362595      Walmart   
7             234  10/21/17  33.65     66999  652377244      Walmart   
8             234  10/16/17  22.30     66999  277394644      Walmart   
9             234  10/20/17  12.32     66999  996849471      Walmart   
10            234  10/22/17  18.25     66999  278354885      Walmart   
11            234  10/23/17  17.35     66999  215480572      Walmart   
12            234  10/21/17   9.53     66999  526069074      Walmart   
13            234  10/24/17   9.94     66999  265077864      Walmart   
14            234  10/18/17  55.21     66999  627831552      Walmart   
15            234  10/17/17  36.82     66999  992105314      Walmart   
16            234  10/16/17  61.56     66999   39408244      Walmart   
17            234  10/20/17   1.34     66999  372268051      Walmart   
18            234  10/28/17  32.01     66999  375916960      Walmart   
19            234  10/21/17  44.76     66999  742726716      Walmart   
20            234  10/18/17   2.65     66999  243131192      Walmart   
21            234  10/25/17  52.15     66999  388544851      Walmart   
22            234  10/29/17  48.86     66999  905914857      Walmart   
23            234  10/28/17   2.22     66999  226080317      Walmart   
24            234  10/21/17  32.20     66999  502816606      Walmart   
25            234  10/24/17  61.77     66999  986134213      Walmart   
26            234  10/18/17  16.64     66999  714264288      Walmart   
27            234  10/18/17  20.64     66999  585091176      Walmart   
28            234  10/28/17  38.82     66999  645118810      Walmart   
29            234  10/17/17  56.14     66999   81313305      Walmart   
...           ...       ...    ...       ...        ...          ...   
11481         304  10/23/17  31.69     98736  430094535  Whole Foods   
11482         304  10/17/17   3.29     98736   17066659  Whole Foods   
11483         304  10/19/17  13.59     98736  634830299  Whole Foods   
11484         304  10/19/17  60.29     98736  295208432  Whole Foods   
11485         304  10/26/17  58.49     98736  319176479  Whole Foods   
11486         304  10/27/17  17.29     98736  303796970  Whole Foods   
11487         304  10/25/17   5.19     98736  236468029  Whole Foods   
11488         304  10/26/17  38.59     98736  212886542  Whole Foods   
11489         304  10/16/17  24.39     98736  795937333  Whole Foods   
11490         304  10/25/17  41.79     98736  709096794  Whole Foods   
11491         304  10/24/17  60.99     98736  162045248  Whole Foods   
11492         304  10/19/17  42.99     98736  312588009  Whole Foods   
11493         304  10/23/17  42.29     98736  514912132  Whole Foods   
11494         304  10/22/17  18.89     98736  978196729  Whole Foods   
11495         304  10/19/17  17.29     98736  418444428  Whole Foods   
11496         304  10/23/17   8.49     98736  529559570  Whole Foods   
11497         304  10/17/17  45.09     98736  804094528  Whole Foods   
11498         304  10/29/17  36.69     98736  556223651  Whole Foods   
11499         304  10/29/17  27.19     98736  388197861  Whole Foods   
11500         304  10/23/17  63.19     98736  165324638  Whole Foods   
11501         304  10/16/17   7.69     98736  827210689  Whole Foods   
11502         304  10/29/17  58.39     98736  506732002  Whole Foods   
11503         304  10/20/17  34.99     98736  545133835  Whole Foods   
11504         304  10/24/17  50.09     98736  3

Now we can use a pivot table to transform our sources into a cross-tabulation of regional prices alongside each other broken down by banner 

In [17]:
final_df = pd.pivot_table(merged, values = 'Price', index = ['Banner', 'UPC'], columns = ['Region'])

Now we can simply write this out to a csv file

In [18]:
final_df.to_csv('regional_prices.csv')

If we simply look at the spreadsheet now, one anamolie becomes quite obvious, so I'll print the final dataframe here to show it

In [19]:
final_df

Region                 Kansas  New York  Northern California  Texas
Banner      UPC                                                    
Safeway     11873171      NaN       NaN                 6.09   5.19
            15052612    53.99       NaN                  NaN  54.49
            16482322    17.89       NaN                  NaN  18.09
            16729338     7.99       NaN                 9.39   8.09
            16829288     3.59       NaN                 4.19   3.59
            16900911    29.19       NaN                34.19  29.49
            16999755     5.29       NaN                 6.19   5.29
            17066659     3.09       NaN                 3.59   3.09
            19696884      NaN       NaN                  NaN   5.89
            19911643      NaN       NaN                  NaN  30.69
            19982129    27.39       NaN                  NaN  27.69
            21872584     7.99       NaN                  NaN    NaN
            22132676     5.39       NaN                  NaN    NaN
            22976162    49.59       NaN                58.19  50.09
            23493525    26.19       NaN                  NaN  26.49
            24764365      NaN       NaN                  NaN  18.29
            25023603    14.99       NaN                  NaN  15.19
            25313327    17.29       NaN                  NaN  17.49
            26132370      NaN       NaN                23.49  20.19
            26791818    10.69       NaN                  NaN  10.79
            26959766    54.39       NaN                63.79  54.99
            27595398     8.59       NaN                  NaN    NaN
            27866264    56.39       NaN                66.19  56.99
            35072231    49.69       NaN                  NaN  50.19
            38126522    57.29       NaN                  NaN  57.89
            38399923    38.79       NaN                45.49  39.19
            39601228    22.69       NaN                  NaN  22.89
            40555797    27.39       NaN                32.19  27.69
            41066545    56.09       NaN                  NaN  56.69
            43366358    46.59       NaN                  NaN  47.09
...                       ...       ...                  ...    ...
Whole Foods 975326196    1.99     30.89                34.79  29.99
            975793603    1.99     45.99                51.89    NaN
            976976235    1.99       NaN                72.49  62.49
            977134983    1.99       NaN                  NaN  34.79
            978164640    1.99       NaN                  NaN    NaN
            978196729    1.99     19.49                21.99  18.89
            979616943    1.99       NaN                  NaN   1.29
            980113553     NaN       NaN                56.49    NaN
            980150388    1.99     56.79                63.99  55.09
            980564443    1.99     38.49                43.39  37.39
            981197816    1.99     48.89                  NaN    NaN
            981510405    1.99       NaN                73.69    NaN
            981620604    1.99     41.39                  NaN  40.19
            986041768    1.99     26.59                29.99    NaN
            986134213    1.99     64.19                72.39  62.29
            987411986    1.99     61.79                  NaN    NaN
            987897892    1.99      9.89                  NaN   9.59
            988776021    1.99       NaN                  NaN    NaN
            988779486    1.99     34.69                39.19    NaN
            990258086    1.99     43.09                48.59    NaN
            992105314    1.99     38.29                  NaN    NaN
            992125856    1.99     49.19                  NaN    NaN
            993910240    1.99     25.89                29.19    NaN
            994344572     NaN     11.89                  NaN  11.59
            995745822     NaN     44.59                  NaN    NaN
            995798889    1.99     62.39                7

What pops out here is that every price reported from the Whole Foods in Kansas is 1.99.  Let's look a little further for anamolies though. To check for outlier prices, we're going to loop through an all unique upcs, creating a smaller dataframe with all price data for a single product. We will then check this array for any entries whose price is more than three standard deviations from the mean price and append it to an outlier dataframe.

In [31]:
all_upc = merged['UPC'].unique()  #create array to loop over
outlier_prices = pd.DataFrame(columns = ['Auditor ID', 'Date', 'Price', 'Store ID', 'UPC', 'Banner', 'Region'])
#create empty dataframe for outliers

In [32]:
for item in all_upc:
    mask = merged['UPC'] ==item
    df = merged[mask]   #create dataframe with all price data for a single upc
    outliers = df[np.abs(df['Price']-df['Price'].mean()) > (3*df['Price'].std())]
    #identify entries with price outside 3 standard deviations of the mean
    outlier_prices = pd.concat([outliers, outlier_prices])  #append outliers to single dataframe

In [35]:
len(merged)

11511

As expected, it looks like we get back all the 1.99 prices from the Whole Foods in Kansas, but let's check if there are any other outliers by grouping the dataframe by price and checking the lenght

In [30]:
len(outlier_prices.groupby('Price'))

2

In [36]:
outlier_prices['Price'].unique() #find out what the other outlier price is

array([ 1.99,  1.49])

In [37]:
outlier_prices.groupby('Price').get_group(1.49) #let's look at the 1.49 price group

Auditor ID      Date  Price  Store ID          UPC       Banner  Region
7320       713.0  10/29/17   1.49   39287.0  835193761.0  Whole Foods  Kansas

Interesting that this one came from the same store and same auditor as well.  Let's see if all the outliers came from the same auditor and store location

In [40]:
len(outlier_prices.groupby(['Banner', 'Region', 'Auditor ID']))

1

Yes, when we group by banner, region, and auditor ID, there is only one group meaning all outliers came from the same auditor at the same store location. Let's take a closer look at this particular auditor by making a dataframe with just their data 

In [41]:
auditor713 = merged.groupby('Auditor ID').get_group(713)

In [42]:
auditor713['Price'].unique()  #did they just price everything 1.99 and one item 1.49?

array([ 51.99,  43.59,  21.39,   1.99,  46.79,  40.99,  26.99,  33.99,
        54.69,  44.09,  54.19,  55.19,  50.29,  23.29,  17.39,   7.79,
        39.49,  53.39,  44.19,   3.19,  46.29,  54.29,  34.39,  49.89,
        53.99,  55.59,  34.99,  56.89,  40.29,  11.29,  21.09,  12.29,
        39.89,  19.69,  41.99,  12.49,  29.19,  49.99,  13.99,  18.89,
        10.39,  31.59,   3.69,  24.09,  18.59,  36.39,  45.59,  41.39,
        12.09,  40.09,  13.09,  36.09,  44.39,  43.19,   1.89,  52.99,
        17.79,  51.89,  52.39,  47.49,  31.39,   4.89,   2.39,  27.09,
        35.89,  56.69,  28.69,  51.09,  11.99,  15.69,  38.99,  44.59,
        11.19,   4.29,  53.69,  10.59,  33.69,  23.69,  47.59,  19.19,
        35.59,  25.79,   2.99,   7.39,  14.49,  37.89,  35.29,  19.39,
        56.79,  36.89,  55.69,  43.39,  42.09,  19.09,  37.99,  11.39,
        38.19,  48.19,  18.49,  40.19,  19.99,   3.49,  32.79,   7.69,
        50.19,  38.59,  27.49,  17.89,   1.09,  13.59,  35.49,  45.39,
      

No, not everything they priced was 1.99 or 1.49.  Let's check if they went to any stores other than the Whole Foods

In [43]:
auditor713['Banner'].unique()

array(['Wegmans', 'Whole Foods'], dtype=object)

Okay, they went to Wegmans and Whole Foods. Let's take a look at the two stores separately

In [44]:
whole_foods = auditor713.groupby('Banner').get_group('Whole Foods')
wegmans = auditor713.groupby('Banner').get_group('Wegmans')

Now that we've got a dataframe with the recorded prices for each store, let's see how many unique prices the auditor recorded from each

In [45]:
whole_foods['Price'].unique()

array([ 1.99,  1.49])

In [48]:
whole_foods.groupby('Price').get_group(1.49)

Auditor ID      Date  Price  Store ID        UPC       Banner  Region
7320         713  10/29/17   1.49     39287  835193761  Whole Foods  Kansas

In [51]:
print(len(wegmans['Price'].unique()))
wegmans.groupby('Price').get_group(1.99) #check if anything was priced 1.99 at wegmans too

296


Auditor ID      Date  Price  Store ID        UPC   Banner  Region
5628         713  10/21/17   1.99      3948  226080317  Wegmans  Kansas

In [50]:
len(whole_foods)

913

All 913 price entries from the Whole Foods are 1.99 plus the one 1.49 price.  Although our quick sweep to check for outliers did not flag all of these, it is probably best not to trust any of the price entries from Auditor 713 for the Whole Foods in Kansas. This auditor's prices from Wegmans however were not flagged as being more than three standard deviations from the mean and are varied enough that they can likely be trusted. In conclusion, I found two things off with the data collected. 
1) Store IDs 38472 and 60957 have prices reported in the 'prices.csv' file, but have no corresponding Banner and Region infor in the file 'stores.json'
2) All price entries from Auditor 713 for the Whole Foods in Kansas should not be trusted as they are all 1.99 with one 1.49